File: go/native/sign.go
package native

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"strconv"
	"time"
)

// signRequest DeviceSign 签名 + 一次性 nonce。
// 与平台 AS `pkg/sign` 完全同构：signData = deviceID + "\n" + timestamp + "\n" + body。
func signRequest(appSecret, deviceID, body string) (SignResult, error) {
	ts := strconv.FormatInt(time.Now().Unix(), 10)
	signData := deviceID + "\n" + ts + "\n" + body

	mac := hmac.New(sha256.New, []byte(appSecret))
	mac.Write([]byte(signData))
	signature := hex.EncodeToString(mac.Sum(nil))

	nonce, err := randomHex(16)
	if err != nil {
		return SignResult{}, err
	}
	return SignResult{Sign: signature, Timestamp: ts, Nonce: nonce}, nil
}

// verifyUpgrade 校验升级清单签名。服务端签名 = HMAC(app_secret, manifestJSON)。
func verifyUpgrade(appSecret, manifestJSON, signature string) (bool, error) {
	if signature == "" {
		return false, nil
	}
	mac := hmac.New(sha256.New, []byte(appSecret))
	mac.Write([]byte(manifestJSON))
	expected := hex.EncodeToString(mac.Sum(nil))
	return hmac.Equal([]byte(expected), []byte(signature)), nil
}

func randomHex(n int) (string, error) {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

File: rust/native/src/sign.rs
//! DeviceSign 签名 + 升级清单校验。
//! 与平台 AS `pkg/sign` 同构：signData = deviceID + "\n" + timestamp + "\n" + body。

use crate::SignResult;
use hmac::{Hmac, Mac};
use sha2::Sha256;

type HmacSha256 = Hmac<Sha256>;

fn hmac_sha256_hex(key: &str, data: &str) -> String {
    let mut mac = HmacSha256::new_from_slice(key.as_bytes()).expect("HMAC key");
    mac.update(data.as_bytes());
    hex::encode(mac.finalize().into_bytes())
}

/// 生成一次性 nonce（32 hex 字符）。
fn random_nonce() -> String {
    use rand::RngCore;
    let mut b = [0u8; 16];
    rand::rngs::OsRng.fill_bytes(&mut b);
    hex::encode(b)
}

pub fn sign_request(
    app_secret: &str,
    device_id: &str,
    body: &str,
) -> Result<SignResult, Box<dyn std::error::Error>> {
    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map_err(|e| e.to_string())?
        .as_secs()
        .to_string();
    let sign_data = format!("{device_id}\n{timestamp}\n{body}");
    let sign = hmac_sha256_hex(app_secret, &sign_data);
    Ok(SignResult {
        sign,
        timestamp,
        nonce: random_nonce(),
    })
}

/// 校验升级清单签名（HMAC(app_secret, manifestJSON)）。
pub fn verify_upgrade(app_secret: &str, manifest_json: &str, signature: &str) -> bool {
    if signature.is_empty() {
        return false;
    }
    let expected = hmac_sha256_hex(app_secret, manifest_json);
    expected == signature
}
